cartographer_ros/cartographer_ros/playable_bag.cc

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer_ros/playable_bag.h"
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer_ros/node_constants.h"
+#include "glog/logging.h"
+#include "tf2_msgs/TFMessage.h"
+
+namespace cartographer_ros {
+
+PlayableBag::PlayableBag(const std::string& bag_filename, const int bag_id,
+                         const ros::Time start_time, const ros::Time end_time,
+                         const ros::Duration buffer_delay,
+                         BufferCallback buffer_callback)
+    : bag_(cartographer::common::make_unique<rosbag::Bag>(
+          bag_filename, rosbag::bagmode::Read)),
+      view_(cartographer::common::make_unique<rosbag::View>(*bag_, start_time,
+                                                            end_time)),
+      view_iterator_(view_->begin()),
+      finished_(false),
+      bag_id_(bag_id),
+      bag_filename_(bag_filename),
+      duration_in_seconds_(
+          (view_->getEndTime() - view_->getBeginTime()).toSec()),
+      log_counter_(0),
+      buffer_delay_(buffer_delay),
+      buffer_callback_(std::move(buffer_callback)) {
+  AdvanceUntilMessageAvailable();
+}
+
+ros::Time PlayableBag::PeekMessageTime() {
+  CHECK(IsMessageAvailable());
+  return buffered_messages_.front().getTime();
+}
+
+std::pair<ros::Time, ros::Time> PlayableBag::GetBeginEndTime() {
+  return std::make_pair(view_->getBeginTime(), view_->getEndTime());
+}
+
+rosbag::MessageInstance PlayableBag::GetNextMessage() {
+  CHECK(IsMessageAvailable());
+  const rosbag::MessageInstance msg = buffered_messages_.front();
+  buffered_messages_.pop_front();
+  AdvanceUntilMessageAvailable();
+  if ((log_counter_++ % 10000) == 0) {
+    LOG(INFO) << "Processed " << (msg.getTime() - view_->getBeginTime()).toSec()
```

Comment:
I thought so at first, but I think it's better to have per-bag progress indicators rather than have different bags race who is going to win the modulo 10000 condition. Let me know if you still think `LOG_EVERY_N` is better.